// @tsed/platform-exceptions/lib/cjs/components/ErrorFilter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorFilter = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const catch_1 = require("../decorators/catch");
let ErrorFilter = class ErrorFilter {
    catch(error, ctx) {
        const { response, logger, env } = ctx;
        const err = this.mapError(error, env);
        response
            .setHeaders(this.getHeaders(error))
            .status(err.status)
            .contentType("application/json")
            .body(env === core_1.Env.PROD ? "InternalServerError" : err);
    }
    mapError(error, env) {
        return {
            name: error.origin?.name || error.name || error.code,
            message: error.message,
            status: error.status || error.statusCode || 500,
            errors: this.getErrors(error),
            stack: env === core_1.Env.DEV ? error.stack : undefined
        };
    }
    getErrors(error) {
        return [error, error.origin].filter(Boolean).reduce((errs, { errors }) => {
            return errs.concat(errors).filter(Boolean);
        }, []);
    }
    getHeaders(error) {
        return [error, error.origin].filter(Boolean).reduce((obj, { headers }) => {
            return {
                ...obj,
                ...(headers || {})
            };
        }, {});
    }
};
ErrorFilter = tslib_1.__decorate([
    (0, catch_1.Catch)(Error)
], ErrorFilter);
exports.ErrorFilter = ErrorFilter;
