// @tsed/common/lib/cjs/decorators/multer/multipartFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipartFile = void 0;
const core_1 = require("@tsed/core");
const platform_params_1 = require("@tsed/platform-params");
const schema_1 = require("@tsed/schema");
const PlatformMulterMiddleware_1 = require("../../middlewares/PlatformMulterMiddleware");
function mapOptions(name, maxCount) {
    return {
        fields: [
            {
                name,
                maxCount
            }
        ]
    };
}
/**
 * Define a parameter as Multipart file.
 *
 * ```typescript
 * import {Controller, Post} from "@tsed/common";
 * import {MulterOptions, MultipartFile} from "@tsed/common";
 * import {Multer} from "@types/multer";
 *
 * type MulterFile = Express.Multer.File;
 *
 * @Controller('/')
 * class MyCtrl {
 *   @Post('/file')
 *   private uploadFile(@MultipartFile("file1") file: MulterFile) {
 *
 *   }
 *
 *   @Post('/file')
 *   @MulterOptions({dest: "/other-dir"})
 *   private uploadFile(@MultipartFile("file1") file: MulterFile) {
 *
 *   }
 *
 *   @Post('/file2')
 *   @MulterOptions({dest: "/other-dir"})
 *   private uploadFile(@MultipartFile("file1") file: MulterFile, @MultipartFile("file2") file2: MulterFile) {
 *
 *   }
 *
 *   @Post('/files')
 *   private uploadFile(@MultipartFile("file1") files: MulterFile[]) {
 *
 *   }
 * }
 * ```
 *
 * > See the tutorial on the [multer configuration](/tutorials/multer.md).
 *
 * @param name
 * @param maxCount
 * @returns Function
 * @decorator
 * @input
 */
function MultipartFile(name, maxCount) {
    return (...args) => {
        const [target, propertyKey, index] = args;
        const multiple = core_1.Metadata.getParamTypes(target, propertyKey)[index] === Array;
        name = (typeof name === "object" ? undefined : name);
        const expression = [name, !multiple && "0"].filter(Boolean).join(".");
        const decorators = (0, core_1.useDecorators)((0, schema_1.InFile)(name), (0, core_1.useMethodDecorators)((0, core_1.StoreMerge)(PlatformMulterMiddleware_1.PlatformMulterMiddleware, mapOptions(name, maxCount))), (0, platform_params_1.UseParam)({
            paramType: platform_params_1.ParamTypes.FILES,
            dataPath: "$ctx.request.files",
            expression,
            useValidation: true
        }));
        decorators(...args);
    };
}
exports.MultipartFile = MultipartFile;
