// @tsed/di/lib/cjs/node/utils/asyncHookContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bindContext = exports.setContext = exports.runInContext = exports.getContext = exports.useContextRef = exports.getAsyncStore = void 0;
const async_hooks_1 = require("async_hooks");
const storage = new async_hooks_1.AsyncLocalStorage();
function getAsyncStore() {
    return storage;
}
exports.getAsyncStore = getAsyncStore;
function useContextRef() {
    return getAsyncStore().getStore();
}
exports.useContextRef = useContextRef;
function getContext() {
    return useContextRef()?.current;
}
exports.getContext = getContext;
async function runInContext(ctx, cb, injector) {
    const ref = useContextRef();
    if (ref) {
        ctx && setContext(ctx);
        return cb();
    }
    else {
        injector = ctx?.injector || injector;
        cb = (await injector?.alterAsync("$alterRunInContext", cb)) || cb;
        return storage.run({ current: ctx }, cb);
    }
}
exports.runInContext = runInContext;
function setContext(ctx) {
    const ref = useContextRef();
    if (ref && !ref.current) {
        ref.current = ctx;
    }
}
exports.setContext = setContext;
function bindContext(cb) {
    return async_hooks_1.AsyncResource.bind(cb);
}
exports.bindContext = bindContext;
