// @tsed/platform-router/lib/esm/domain/PlatformLayer.js
import { isPrimitive, nameOf } from "@tsed/core";
import { concatPath } from "@tsed/schema";
export class PlatformLayer {
    provider;
    path = "";
    method;
    handlers = [];
    router;
    opts = {};
    layers = [];
    parent;
    basePath;
    #args;
    constructor(props = {}) {
        Object.assign(this, props);
        this.layers = this.layers.map((layer) => {
            return new PlatformLayer({
                ...layer,
                parent: this
            });
        });
    }
    set(args) {
        this.#args = args;
    }
    getArgs() {
        return [this.path, ...this.#args].filter(Boolean);
    }
    isProvider() {
        return !!this.provider;
    }
    inspect() {
        return {
            basePath: this.getBasePath(),
            path: this.path,
            method: this.method,
            handlers: this.handlers.map((item) => String(item)),
            opts: Object.entries(this.opts).reduce((obj, [key, value]) => {
                return {
                    ...obj,
                    [key]: isPrimitive(value) ? value : nameOf(value)
                };
            }, {})
        };
    }
    getBasePath() {
        return concatPath(this.parent?.getBasePath(), this.basePath);
    }
    setLayers(layers) {
        this.layers = layers.map((layer) => {
            return new PlatformLayer({
                ...layer,
                parent: this
            });
        });
    }
}
