// @tsed/platform-router/lib/esm/domain/PlatformHandlerMetadata.js
import { nameOf } from "@tsed/core";
import { ProviderScope } from "@tsed/di";
import { ParamTypes } from "@tsed/platform-params";
import { EndpointMetadata, JsonEntityStore, JsonParameterStore } from "@tsed/schema";
import { PlatformHandlerType } from "./PlatformHandlerType.js";
export class PlatformHandlerMetadata {
    path;
    provider;
    propertyKey;
    type = PlatformHandlerType.RAW_FN;
    hasNextFunction = false;
    opts = {};
    compiledHandler;
    #handler;
    constructor(props) {
        const { propertyKey, type, provider, handler, opts } = props;
        this.provider = provider;
        this.type = type || handler.type || PlatformHandlerType.RAW_FN;
        this.opts = opts || {};
        this.#handler = propertyKey ? this.target.prototype[propertyKey] : handler;
        if (propertyKey) {
            this.propertyKey = propertyKey;
            this.hasNextFunction = this.hasParamType(ParamTypes.NEXT_FN);
            if (this.hasParamType(ParamTypes.ERR)) {
                this.type = PlatformHandlerType.ERR_MIDDLEWARE;
            }
        }
        else {
            if (this.#handler.length === 4) {
                this.type = PlatformHandlerType.RAW_ERR_FN;
            }
            this.hasNextFunction = this.#handler.length >= 3;
        }
    }
    get target() {
        return this.provider?.useClass || this.#handler;
    }
    get token() {
        return this.provider?.token || this.#handler;
    }
    get handler() {
        return this.#handler;
    }
    get scope() {
        return this.provider?.scope || ProviderScope.SINGLETON;
    }
    get hasErrorParam() {
        return this.type === PlatformHandlerType.ERR_MIDDLEWARE || this.type === PlatformHandlerType.RAW_ERR_FN;
    }
    get store() {
        return JsonEntityStore.fromMethod(this.provider.useClass, this.propertyKey);
    }
    static from(injector, input, opts = {}) {
        if (input instanceof PlatformHandlerMetadata) {
            return input;
        }
        if (input instanceof EndpointMetadata) {
            const provider = injector.getProvider(opts.token);
            return new PlatformHandlerMetadata({
                provider,
                type: PlatformHandlerType.ENDPOINT,
                propertyKey: input.propertyKey,
                opts
            });
        }
        const provider = injector.getProvider(input);
        if (provider) {
            return new PlatformHandlerMetadata({
                provider,
                type: PlatformHandlerType.MIDDLEWARE,
                propertyKey: "use",
                opts
            });
        }
        return new PlatformHandlerMetadata({
            handler: input,
            type: input.type,
            opts
        });
    }
    getParams() {
        return JsonParameterStore.getParams(this.target, this.propertyKey) || [];
    }
    hasParamType(paramType) {
        return this.getParams().findIndex((p) => p.paramType === paramType) > -1;
    }
    isInjectable() {
        return !(this.isRawFn() || this.isResponseFn());
    }
    isRawFn() {
        return this.type === PlatformHandlerType.RAW_ERR_FN || this.type === PlatformHandlerType.RAW_FN;
    }
    isEndpoint() {
        return this.type === PlatformHandlerType.ENDPOINT;
    }
    isCtxFn() {
        return this.type === PlatformHandlerType.CTX_FN;
    }
    isResponseFn() {
        return this.type === PlatformHandlerType.RESPONSE_FN;
    }
    toString() {
        return [nameOf(this.target), this.propertyKey].filter(Boolean).join(".");
    }
}
