// @tsed/platform-router/lib/cjs/domain/PlatformLayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformLayer = void 0;
const core_1 = require("@tsed/core");
const schema_1 = require("@tsed/schema");
class PlatformLayer {
    provider;
    path = "";
    method;
    handlers = [];
    router;
    opts = {};
    layers = [];
    parent;
    basePath;
    #args;
    constructor(props = {}) {
        Object.assign(this, props);
        this.layers = this.layers.map((layer) => {
            return new PlatformLayer({
                ...layer,
                parent: this
            });
        });
    }
    set(args) {
        this.#args = args;
    }
    getArgs() {
        return [this.path, ...this.#args].filter(Boolean);
    }
    isProvider() {
        return !!this.provider;
    }
    inspect() {
        return {
            basePath: this.getBasePath(),
            path: this.path,
            method: this.method,
            handlers: this.handlers.map((item) => String(item)),
            opts: Object.entries(this.opts).reduce((obj, [key, value]) => {
                return {
                    ...obj,
                    [key]: (0, core_1.isPrimitive)(value) ? value : (0, core_1.nameOf)(value)
                };
            }, {})
        };
    }
    getBasePath() {
        return (0, schema_1.concatPath)(this.parent?.getBasePath(), this.basePath);
    }
    setLayers(layers) {
        this.layers = layers.map((layer) => {
            return new PlatformLayer({
                ...layer,
                parent: this
            });
        });
    }
}
exports.PlatformLayer = PlatformLayer;
