// @tsed/platform-params/lib/esm/domain/ParamTypes.js
export var ParamTypes;
(function (ParamTypes) {
    ParamTypes["$CTX"] = "$CTX";
    ParamTypes["BODY"] = "BODY";
    ParamTypes["PATH"] = "PATH";
    ParamTypes["QUERY"] = "QUERY";
    ParamTypes["HEADER"] = "HEADER";
    ParamTypes["RAW_BODY"] = "RAW_BODY";
    ParamTypes["COOKIES"] = "COOKIES";
    ParamTypes["LOCALS"] = "LOCALS";
    ParamTypes["SESSION"] = "SESSION";
    ParamTypes["FILES"] = "FILES";
    /// NOT STD
    ParamTypes["NEXT_FN"] = "NEXT_FN";
    ParamTypes["ERR"] = "ERR";
    ParamTypes["REQUEST"] = "REQUEST";
    ParamTypes["PLATFORM_REQUEST"] = "PLATFORM_REQUEST";
    ParamTypes["NODE_REQUEST"] = "NODE_REQUEST";
    ParamTypes["RESPONSE"] = "RESPONSE";
    ParamTypes["PLATFORM_RESPONSE"] = "PLATFORM_RESPONSE";
    ParamTypes["NODE_RESPONSE"] = "NODE_RESPONSE";
    ParamTypes["RESPONSE_DATA"] = "RESPONSE_DATA";
    ParamTypes["ENDPOINT_INFO"] = "ENDPOINT_INFO";
    ParamTypes["RES"] = "RES";
    ParamTypes["REQ"] = "REQ";
})(ParamTypes || (ParamTypes = {}));
export const PARAM_TYPES_DATA_PATH = {
    [ParamTypes.HEADER]: "$ctx.request.headers",
    [ParamTypes.RAW_BODY]: "$ctx.request.rawBody",
    [ParamTypes.BODY]: "$ctx.request.body",
    [ParamTypes.QUERY]: "$ctx.request.query",
    [ParamTypes.PATH]: "$ctx.request.params",
    [ParamTypes.COOKIES]: "$ctx.request.cookies",
    [ParamTypes.SESSION]: "$ctx.request.session",
    [ParamTypes.LOCALS]: "$ctx.response.locals"
};
