// @tsed/schema/lib/cjs/domain/JsonDiscriminator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Discriminator = void 0;
const core_1 = require("@tsed/core");
class Discriminator {
    propertyName;
    base;
    values = new Map();
    types = new Map();
    constructor({ base, propertyName, types, values } = {}) {
        Object.assign(this, (0, core_1.cleanObject)({
            base,
            propertyName,
            types: types ? new Map(types) : undefined,
            values: values ? new Map(values) : undefined
        }));
    }
    add(type, values) {
        values.forEach((value) => {
            this.values.set(value, type);
        });
        this.types.set(type, values);
        return this;
    }
    getType(discriminatorValue) {
        return this.values.get(discriminatorValue) || this.base;
    }
    getValues(type) {
        return this.types.get(type);
    }
    getDefaultValue(type) {
        const values = this.types.get(type);
        return values ? values[0] : undefined;
    }
    children() {
        return [...new Set([...this.types.keys()])];
    }
}
exports.Discriminator = Discriminator;
