// @tsed/schema/lib/cjs/domain/JsonPropertyStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PropertyMetadata = exports.JsonPropertyStore = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const JsonEntityStore_1 = require("./JsonEntityStore");
const JsonSchema_1 = require("./JsonSchema");
const jsonEntityComponent_1 = require("../decorators/config/jsonEntityComponent");
let JsonPropertyStore = class JsonPropertyStore extends JsonEntityStore_1.JsonEntityStore {
    parent = JsonEntityStore_1.JsonEntityStore.from(this.target);
    /**
     * Return the required state.
     * @returns {boolean}
     */
    get required() {
        return this.parent.schema.isRequired(this.propertyKey);
    }
    /**
     * Change the state of the required data.
     * @param value
     */
    set required(value) {
        if (value) {
            this.parent.schema.addRequired(this.propertyKey);
        }
        else {
            this.parent.schema.removeRequired(this.propertyKey);
        }
    }
    get allowedRequiredValues() {
        return this.schema.$allow;
    }
    discriminatorKey() {
        this.parent.schema.discriminatorKey(String(this.propertyKey));
        this.itemSchema.isDiscriminatorKey = true;
        return this;
    }
    isDiscriminatorKey() {
        return this.itemSchema.isDiscriminatorKey;
    }
    /**
     * Check precondition between value, required and allowedRequiredValues to know if the entity is required.
     * @param value
     * @returns {boolean}
     */
    isRequired(value) {
        return this.required && [undefined, null, ""].includes(value) && !this.allowedRequiredValues.includes(value);
    }
    build() {
        if (!this._type) {
            this.buildType(core_1.Metadata.getType((0, core_1.prototypeOf)(this.target), this.propertyKey));
        }
        this._type = this._type || Object;
        const properties = this.parent.schema.get("properties");
        let schema = properties[this.propertyName];
        if (!schema) {
            this.parent.children.set(this.propertyName, this);
            schema = JsonSchema_1.JsonSchema.from({
                type: this.collectionType || this.type
            });
            if (this.collectionType) {
                schema.itemSchema(this.type);
            }
        }
        this.parent.schema.addProperty(this.propertyName, schema);
        this._schema = schema;
    }
    static get(target, propertyKey) {
        return JsonEntityStore_1.JsonEntityStore.from((0, core_1.prototypeOf)(target), propertyKey);
    }
};
JsonPropertyStore = tslib_1.__decorate([
    (0, jsonEntityComponent_1.JsonEntityComponent)(core_1.DecoratorTypes.PROP)
], JsonPropertyStore);
exports.JsonPropertyStore = JsonPropertyStore;
exports.PropertyMetadata = JsonPropertyStore;
