// @tsed/engines/lib/cjs/components/ReactEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactEngine = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
const fs_1 = require("fs");
const viewEngine_1 = require("../decorators/viewEngine");
const Engine_1 = require("./Engine");
const cache_1 = require("../utils/cache");
let ReactEngine = class ReactEngine extends Engine_1.Engine {
    constructor(name, options) {
        super(name, options);
        this.patchRequire();
    }
    async $onInit() {
        await super.$onInit();
        await (0, cache_1.importEngine)("react-dom/server", "ReactDOM");
        await (0, cache_1.importEngine)("@babel/core", "babel");
    }
    get babel() {
        return (0, cache_1.getCachedEngine)("babel");
    }
    get reactDOM() {
        return (0, cache_1.getCachedEngine)("ReactDOM");
    }
    patchRequire() {
        if (require.extensions) {
            // Ensure JSX is transformed on require
            if (!require.extensions[".jsx"]) {
                require.extensions[".jsx"] = this.requireReact.bind(this);
            }
            // Supporting .react extension as well as test cases
            // Using .react extension is not recommended.
            if (!require.extensions[".react"]) {
                require.extensions[".react"] = this.requireReact.bind(this);
            }
        }
    }
    configure(options) {
        const base = options.base;
        delete options.base;
        const enableCache = options.cache;
        delete options.cache;
        const isNonStatic = options.isNonStatic;
        delete options.isNonStatic;
        return { base, enableCache, isNonStatic };
    }
    $compile(template, options) {
        // Assign HTML Base
        const { base, enableCache, isNonStatic } = this.configure(options);
        // Start Conversion
        const Code = this.requireReactString(template);
        const Factory = this.engine.createFactory(Code);
        return () => {
            const parsed = new Factory(options);
            const content = isNonStatic ? this.reactDOM.renderToString(parsed) : this.reactDOM.renderToStaticMarkup(parsed);
            if (base) {
                options.content = content;
                return this.renderWithBase(template, base, enableCache, options);
            }
            return content;
        };
    }
    async $compileFile(file, options) {
        var _a;
        // Assign HTML Base
        const { base, enableCache, isNonStatic } = this.configure(options);
        let path = (0, path_1.resolve)(file);
        delete require.cache[path];
        const ReactDOM = await (0, cache_1.importEngine)("react-dom/server", "ReactDOM");
        const { default: Code } = await (_a = path, Promise.resolve().then(() => tslib_1.__importStar(require(_a))));
        const Factory = this.engine.createFactory(Code);
        return () => {
            const parsed = new Factory(options);
            const content = isNonStatic ? ReactDOM.renderToString(parsed) : ReactDOM.renderToStaticMarkup(parsed);
            if (base) {
                options.content = content;
                return this.renderWithBase(file, base, enableCache, options);
            }
            return content;
        };
    }
    renderWithBase(key, base, enableCache, options) {
        const baseStr = (0, cache_1.getFromCache)(key) || (0, fs_1.readFileSync)((0, path_1.resolve)(base), "utf8");
        if (enableCache) {
            (0, cache_1.setToCache)(key, baseStr);
        }
        return this.reactBaseTmpl(baseStr, options);
    }
    transformFileSync(filename) {
        return this.babel.transformFileSync(filename, { presets: ["@babel/preset-react"] }).code;
    }
    transform(src) {
        return this.babel.transform(src, { presets: ["@babel/preset-react"] }).code;
    }
    requireReact(module, filename) {
        const compiled = this.transformFileSync(filename);
        return module._compile(compiled, filename);
    }
    requireReactString(src, filename) {
        if (!filename)
            filename = "";
        // @ts-ignore
        const m = new module.constructor();
        filename = filename || "";
        // Compile Using React
        const compiled = this.transform(src);
        // Compile as a module
        m.paths = module.paths;
        m._compile(compiled, filename);
        return m.exports;
    }
    reactBaseTmpl(data, options) {
        let exp;
        let regex;
        // Iterates through the keys in file object
        // and interpolate / replace {{key}} with it's value
        for (const k in options) {
            if (options.hasOwnProperty(k)) {
                exp = `{{${k}}}`;
                regex = new RegExp(exp, "g");
                if (data.match(regex)) {
                    data = data.replace(regex, options[k]);
                }
            }
        }
        return data;
    }
};
ReactEngine = tslib_1.__decorate([
    (0, viewEngine_1.ViewEngine)("react", {
        requires: ["react"]
    }),
    tslib_1.__metadata("design:paramtypes", [String, Object])
], ReactEngine);
exports.ReactEngine = ReactEngine;
