// @tsed/schema/lib/esm/constants/OperationVerbs.js
export var OperationVerbs;
(function (OperationVerbs) {
    OperationVerbs["ALL"] = "ALL";
    OperationVerbs["GET"] = "GET";
    OperationVerbs["POST"] = "POST";
    OperationVerbs["PUT"] = "PUT";
    OperationVerbs["PATCH"] = "PATCH";
    OperationVerbs["HEAD"] = "HEAD";
    OperationVerbs["DELETE"] = "DELETE";
    OperationVerbs["OPTIONS"] = "OPTIONS";
    OperationVerbs["TRACE"] = "TRACE";
    OperationVerbs["PUBLISH"] = "PUBLISH";
    OperationVerbs["SUBSCRIBE"] = "SUBSCRIBE";
    OperationVerbs["CUSTOM"] = "CUSTOM";
})(OperationVerbs || (OperationVerbs = {}));
export const OPERATION_HTTP_VERBS = [
    OperationVerbs.ALL,
    OperationVerbs.GET,
    OperationVerbs.POST,
    OperationVerbs.PUT,
    OperationVerbs.PATCH,
    OperationVerbs.HEAD,
    OperationVerbs.DELETE,
    OperationVerbs.OPTIONS,
    OperationVerbs.TRACE,
    OperationVerbs.CUSTOM
];
export const OPERATION_WS_VERBS = [OperationVerbs.PUBLISH, OperationVerbs.SUBSCRIBE];
/**
 * @deprecated Use OperationVerbs instead of OperationMethods
 */
export const OperationMethods = OperationVerbs;
