// @tsed/platform-views/lib/esm/services/PlatformViews.js
import { __decorate, __metadata } from "tslib";
import { Env, getValue } from "@tsed/core";
import { Constant, Inject, InjectorService, Module } from "@tsed/di";
import { engines, getEngine, requires } from "@tsed/engines";
import Fs from "fs";
import { extname, join, resolve } from "path";
import { PLATFORM_VIEWS_EXTENSIONS } from "../domain/PlatformViewsSettings.js";
async function patchEJS(ejs) {
    if (!ejs) {
        const mod = await import("ejs");
        ejs = mod.default;
    }
    return {
        compile(str, { client, ...options }) {
            return ejs.compile(str, options);
        }
    };
}
/**
 * @platform
 */
let PlatformViews = class PlatformViews {
    env;
    root;
    cache;
    disabled;
    viewEngine;
    extensionsOptions;
    engineOptions;
    injector;
    #extensions;
    #engines = new Map();
    #cachePaths = new Map();
    async $onInit() {
        if (!this.disabled) {
            this.#extensions = new Map(Object.entries({
                ...PLATFORM_VIEWS_EXTENSIONS,
                ...this.extensionsOptions
            }));
            await this.loadEngines();
        }
    }
    async loadEngines() {
        requires.set("ejs", await patchEJS(requires.get("ejs")));
        this.#extensions.forEach((engineType) => {
            if (engines.has(engineType)) {
                const options = this.getEngineOptions(engineType);
                if (options.requires) {
                    requires.set(engineType, options.requires);
                }
                this.registerEngine(engineType, {
                    options,
                    render: getEngine(engineType)
                });
            }
        });
    }
    getEngines() {
        return [...this.#extensions.entries()].map(([extension, engineType]) => {
            const engine = this.getEngine(this.getExtension(engineType));
            return {
                extension,
                engine
            };
        });
    }
    registerEngine(engineType, engine) {
        this.#engines.set(engineType, engine);
        return this;
    }
    getEngine(type) {
        return this.#engines.get(this.#extensions.get(type) || type);
    }
    getEngineOptions(engineType) {
        return getValue(this.engineOptions, engineType, {});
    }
    async render(viewPath, options = {}) {
        const { $ctx } = options;
        options = await this.injector.alterAsync("$alterRenderOptions", options, $ctx);
        const { path, extension } = this.#cachePaths.get(viewPath) || this.#cachePaths.set(viewPath, this.resolve(viewPath)).get(viewPath);
        const engine = this.getEngine(extension);
        if (!engine) {
            throw new Error(`Engine not found to render the following "${viewPath}"`);
        }
        const finalOpts = Object.assign({ cache: this.cache || this.env === Env.PROD }, engine.options, options, { $ctx });
        return engine.render(path, finalOpts);
    }
    getExtension(viewPath) {
        return (extname(viewPath) || this.viewEngine).replace(/\./, "");
    }
    resolve(viewPath) {
        const extension = this.getExtension(viewPath);
        viewPath = viewPath.replace(extname(viewPath), "") + "." + extension;
        const path = [
            viewPath,
            resolve(join(this.root, viewPath)),
            resolve(join(process.cwd(), "views", viewPath)),
            resolve(join(process.cwd(), "public", viewPath))
        ].find((file) => Fs.existsSync(file)) || viewPath;
        return {
            path,
            extension
        };
    }
};
__decorate([
    Constant("env"),
    __metadata("design:type", String)
], PlatformViews.prototype, "env", void 0);
__decorate([
    Constant("views.root", `${process.cwd()}/views`),
    __metadata("design:type", String)
], PlatformViews.prototype, "root", void 0);
__decorate([
    Constant("views.cache"),
    __metadata("design:type", Boolean)
], PlatformViews.prototype, "cache", void 0);
__decorate([
    Constant("views.disabled", false),
    __metadata("design:type", String)
], PlatformViews.prototype, "disabled", void 0);
__decorate([
    Constant("views.viewEngine", "ejs"),
    __metadata("design:type", String)
], PlatformViews.prototype, "viewEngine", void 0);
__decorate([
    Constant("views.extensions", {}),
    __metadata("design:type", Object)
], PlatformViews.prototype, "extensionsOptions", void 0);
__decorate([
    Constant("views.options", {}),
    __metadata("design:type", Object)
], PlatformViews.prototype, "engineOptions", void 0);
__decorate([
    Inject(),
    __metadata("design:type", InjectorService)
], PlatformViews.prototype, "injector", void 0);
PlatformViews = __decorate([
    Module({
        views: {
            exists: true
        }
    })
], PlatformViews);
export { PlatformViews };
