// @tsed/platform-params/lib/cjs/errors/RequiredValidationError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequiredValidationError = void 0;
const core_1 = require("@tsed/core");
const ValidationError_1 = require("./ValidationError");
class RequiredValidationError extends ValidationError_1.ValidationError {
    name = "REQUIRED_VALIDATION_ERROR";
    static from(metadata) {
        const name = (0, core_1.nameOf)(metadata.paramType);
        const expression = metadata.expression;
        const type = name.toLowerCase().replace(/parse|params|filter/gi, "");
        const message = `It should have required parameter '${expression}'`;
        const errors = [
            {
                dataPath: "",
                keyword: "required",
                message,
                modelName: type,
                params: {
                    missingProperty: expression
                },
                schemaPath: "#/required"
            }
        ];
        return new RequiredValidationError(message, errors);
    }
}
exports.RequiredValidationError = RequiredValidationError;
