// @tsed/schema/lib/cjs/utils/getSpec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSpec = void 0;
const core_1 = require("@tsed/core");
const SpecTypes_1 = require("../domain/SpecTypes");
const JsonSchemaMapperContainer_1 = require("../registries/JsonSchemaMapperContainer");
const getJsonEntityStore_1 = require("./getJsonEntityStore");
const mergeSpec_1 = require("./mergeSpec");
const operationIdFormatter_1 = require("./operationIdFormatter");
/**
 * @ignore
 */
const caches = new Map();
/**
 * @ignore
 */
function get(model, options, cb) {
    if (!caches.has(model)) {
        caches.set(model, new Map());
    }
    const cache = caches.get(model);
    const key = JSON.stringify(options);
    if (!cache.has(key)) {
        cache.set(key, cb());
    }
    return cache.get(key);
}
function generate(model, options) {
    const store = (0, getJsonEntityStore_1.getJsonEntityStore)(model);
    const { rootPath = "/" } = options;
    options = {
        ...options,
        rootPath,
        defaultTags: [
            (0, core_1.cleanObject)({
                name: store.schema.getName(),
                description: store.schema.get("description")
            })
        ]
    };
    return (0, JsonSchemaMapperContainer_1.execMapper)("generate", [store], options);
}
/**
 * Return the swagger or open spec for the given class.
 * @param model
 * @param options
 */
function getSpec(model, options = {}) {
    options = {
        specType: SpecTypes_1.SpecTypes.OPENAPI,
        ...options,
        tags: [],
        paths: {},
        channels: {},
        components: {},
        operationIdFormatter: options.operationIdFormatter || (0, operationIdFormatter_1.operationIdFormatter)(options.operationIdPattern),
        root: false
    };
    if ((0, core_1.isArray)(model)) {
        let finalSpec = {};
        options = {
            ...options,
            append(spec) {
                finalSpec = (0, mergeSpec_1.mergeSpec)(finalSpec, spec);
            }
        };
        model.forEach(({ token, ...opts }) => {
            const spec = getSpec(token, {
                ...options,
                ...opts
            });
            options.append(spec);
        });
        return finalSpec;
    }
    return get(model, options, () => generate(model, options));
}
exports.getSpec = getSpec;
