// @tsed/schema/lib/cjs/utils/mergeSpec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeSpec = exports.tagsReducer = exports.security = exports.parameters = exports.schemesReducer = void 0;
const core_1 = require("@tsed/core");
/**
 * @ignore
 */
exports.schemesReducer = (0, core_1.mergeReducerBuilder)((current, value) => (current.type && current.type === value.type) || (current.$ref && current.$ref === value.$ref));
/**
 * @ignore
 */
exports.parameters = (0, core_1.mergeReducerBuilder)((current, value) => current.in === value.in && current.name === value.name);
/**
 * @ignore
 */
exports.security = (0, core_1.mergeReducerBuilder)((current, value) => {
    return !!Object.keys(value).find((key) => !!current[key]);
});
/**
 * @ignore
 */
exports.tagsReducer = (0, core_1.mergeReducerBuilder)((current, value) => current.name === value.name);
/**
 * @ignore
 */
const SPEC_REDUCERS = {
    security: exports.security,
    parameters: exports.parameters,
    oneOf: exports.schemesReducer,
    anyOf: exports.schemesReducer,
    allOf: exports.schemesReducer,
    tags: exports.tagsReducer
};
/**
 * Merge two spec
 * @param spec
 * @param input
 */
function mergeSpec(spec, input) {
    return (0, core_1.deepMerge)(spec, input, {
        reducers: SPEC_REDUCERS,
        cleanUndefinedProps: true
    });
}
exports.mergeSpec = mergeSpec;
