// @tsed/core/lib/cjs/utils/objects/deepMerge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deepMerge = exports.mergeReducerBuilder = void 0;
const createInstance_1 = require("./createInstance");
const isFunction_1 = require("./isFunction");
const isPrimitive_1 = require("./isPrimitive");
const isSymbol_1 = require("./isSymbol");
const objectKeys_1 = require("./objectKeys");
function mergeReducerBuilder(cb) {
    return (collection, value, options) => {
        const index = collection.findIndex((item) => cb(item, value));
        if (index === -1) {
            return [...collection, value];
        }
        collection[index] = deepMerge(collection[index], value, options);
        return collection;
    };
}
exports.mergeReducerBuilder = mergeReducerBuilder;
const defaultReducer = mergeReducerBuilder((a, b) => a === b);
function getReducer({ reducers, parentKey }) {
    if (!reducers) {
        return defaultReducer;
    }
    if (parentKey && reducers[parentKey]) {
        return reducers[parentKey];
    }
    return reducers["default"] || defaultReducer;
}
function shouldReturnObj(obj, source) {
    return (0, isPrimitive_1.isPrimitive)(obj) || (0, isSymbol_1.isSymbol)(obj) || (0, isFunction_1.isFunction)(obj) || source === undefined;
}
function shouldReturnSource(obj, source) {
    return obj === undefined || obj === null || (obj === "" && source !== "");
}
function deepMerge(source, obj, options = {}) {
    if (shouldReturnSource(obj, source)) {
        return source;
    }
    if (shouldReturnObj(obj, source)) {
        return obj;
    }
    if (Array.isArray(source)) {
        const reducer = getReducer(options);
        return [].concat(obj).reduce((out, value) => reducer(out, value, options), [...source]);
    }
    return [...(0, objectKeys_1.objectKeys)(source), ...(0, objectKeys_1.objectKeys)(obj)].reduce((out, key) => {
        const src = source && source[key];
        const value = deepMerge(src, obj && obj[key], {
            ...options,
            parentKey: key
        });
        if (options.cleanUndefinedProps && value === undefined) {
            return out;
        }
        return {
            ...out,
            [key]: value
        };
    }, (0, createInstance_1.createInstance)(source));
}
exports.deepMerge = deepMerge;
