// @tsed/di/lib/cjs/node/services/DITest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DITest = void 0;
const core_1 = require("@tsed/core");
const logger_1 = require("@tsed/logger");
const index_1 = require("../../common/index");
const DIContext_1 = require("../domain/DIContext");
const setLoggerConfiguration_1 = require("../utils/setLoggerConfiguration");
/**
 * Tool to run test with lightweight DI sandbox.
 */
class DITest {
    static options = {};
    static _injector = null;
    static get injector() {
        if (DITest._injector) {
            return DITest._injector;
        }
        /* istanbul ignore next */
        throw new Error("PlatformTest.injector is not initialized. Use PlatformTest.create(): Promise before PlatformTest.invoke() or PlatformTest.injector.\n" +
            "Example:\n" +
            "before(async () => {\n" +
            "   await PlatformTest.create()\n" +
            "   await PlatformTest.invoke(MyService, [])\n" +
            "})");
    }
    static set injector(injector) {
        DITest._injector = injector;
    }
    static set(key, value) {
        (0, core_1.setValue)(DITest.options, key, value);
    }
    static hasInjector() {
        return !!DITest._injector;
    }
    static async create(settings = {}) {
        settings = {
            ...DITest.options,
            ...settings
        };
        DITest.injector = DITest.createInjector(settings);
        await DITest.createContainer();
    }
    static async createContainer() {
        await DITest.injector.load((0, index_1.createContainer)());
    }
    /**
     * Create a new injector with the right default services
     */
    static createInjector(settings = {}) {
        const injector = new index_1.InjectorService();
        injector.logger = logger_1.$log;
        // @ts-ignore
        injector.settings.set(DITest.configure(settings));
        (0, setLoggerConfiguration_1.setLoggerConfiguration)(injector);
        return injector;
    }
    /**
     * Resets the test injector of the test context, so it won't pollute your next test. Call this in your `tearDown` logic.
     */
    static async reset() {
        if (DITest.hasInjector()) {
            await DITest.injector.destroy();
            DITest._injector = null;
        }
    }
    /**
     * Invoke a provider and return a fresh instance
     * @param target
     * @param providers
     */
    static async invoke(target, providers = []) {
        const locals = new index_1.LocalsContainer();
        providers.forEach((p) => {
            locals.set(p.token, p.use);
        });
        locals.set(index_1.InjectorService, DITest.injector);
        const instance = DITest.injector.invoke(target, locals, { rebuild: true });
        if (instance && instance.$onInit) {
            // await instance.$onInit();
            const result = instance.$onInit();
            if (result instanceof Promise) {
                return result.then(() => instance);
            }
        }
        if ((0, core_1.isClass)(instance)) {
            await Promise.all(Object.values(instance).filter(core_1.isPromise));
        }
        return instance;
    }
    /**
     * Return the instance from injector registry
     * @param target
     * @param options
     */
    static get(target, options = {}) {
        return DITest.injector.get(target, options);
    }
    static createDIContext() {
        return new DIContext_1.DIContext({
            id: "id",
            injector: DITest.injector,
            logger: DITest.injector.logger
        });
    }
    static configure(settings = {}) {
        return {
            ...settings,
            env: (0, core_1.getValue)(settings, "env", core_1.Env.TEST),
            logger: {
                ...(0, core_1.getValue)(settings, "logger", {}),
                level: (0, core_1.getValue)(settings, "logger.level", "off")
            }
        };
    }
}
exports.DITest = DITest;
