// @tsed/di/lib/cjs/common/domain/ControllerProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControllerProvider = void 0;
const Provider_1 = require("./Provider");
const ProviderType_1 = require("./ProviderType");
class ControllerProvider extends Provider_1.Provider {
    tokenRouter;
    constructor(provide, options = {}) {
        super(provide, options);
        this.type = ProviderType_1.ProviderType.CONTROLLER;
    }
    /**
     *
     * @returns {any[]}
     */
    get middlewares() {
        return Object.assign({
            use: [],
            useAfter: [],
            useBefore: []
        }, this.store.get("middlewares", {}));
    }
    /**
     *
     * @param middlewares
     */
    set middlewares(middlewares) {
        const mdlwrs = this.middlewares;
        const concat = (key, a, b) => (a[key] = a[key].concat(b[key]));
        Object.keys(middlewares).forEach((key) => {
            concat(key, mdlwrs, middlewares);
        });
        this.store.set("middlewares", mdlwrs);
    }
}
exports.ControllerProvider = ControllerProvider;
