// @tsed/common/lib/cjs/utils/createContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createContext = exports.buildIgnoreLog = void 0;
const uuid_1 = require("uuid");
const PlatformContext_1 = require("../domain/PlatformContext");
const PlatformRequest_1 = require("../services/PlatformRequest");
const PlatformResponse_1 = require("../services/PlatformResponse");
function defaultReqIdBuilder(req) {
    return req.get("x-request-id") || (0, uuid_1.v4)().split("-").join("");
}
function mapIgnoreUrlPatterns(ignoreUrlPatterns) {
    return ignoreUrlPatterns.map((pattern) => (typeof pattern === "string" ? new RegExp(pattern, "gi") : pattern));
}
function buildIgnoreLog(ignoreUrlPatterns) {
    if (ignoreUrlPatterns) {
        ignoreUrlPatterns = mapIgnoreUrlPatterns(ignoreUrlPatterns);
        return (ignore, data, url) => ignoreUrlPatterns?.find((reg) => !!url.match(reg));
    }
}
exports.buildIgnoreLog = buildIgnoreLog;
/**
 * Create the TsED context to wrap request, response, injector, etc...
 * @param injector
 * @ignore
 */
function createContext(injector) {
    const ResponseKlass = injector.getProvider(PlatformResponse_1.PlatformResponse)?.useClass;
    const RequestKlass = injector.getProvider(PlatformRequest_1.PlatformRequest)?.useClass;
    const { reqIdBuilder = defaultReqIdBuilder, ...loggerOptions } = injector.settings.logger;
    const opts = {
        ...loggerOptions,
        logger: injector.logger,
        injector,
        ResponseKlass,
        RequestKlass
    };
    const ignoreLog = buildIgnoreLog(loggerOptions.ignoreUrlPatterns);
    return function invokeContext(event) {
        const $ctx = new PlatformContext_1.PlatformContext({
            ...opts,
            event,
            id: reqIdBuilder(event.request)
        });
        ignoreLog && $ctx.logger.alterIgnoreLog((ignore, data) => ignoreLog(ignore, data, $ctx.url));
        return $ctx;
    };
}
exports.createContext = createContext;
