// @tsed/common/lib/cjs/utils/getConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfiguration = void 0;
const core_1 = require("@tsed/core");
const di_1 = require("@tsed/di");
function getConfiguration(configuration = {}, module = null) {
    if (configuration.$$resolved) {
        return configuration;
    }
    const store = module ? core_1.Store.from(module).get("configuration") || {} : {};
    const env = configuration.env || store.env || process.env.NODE_ENV || core_1.Env.DEV;
    return {
        $$resolved: true,
        env,
        httpPort: 8080,
        httpsPort: false,
        ...store,
        ...configuration,
        scopes: {
            [di_1.ProviderType.CONTROLLER]: di_1.ProviderScope.SINGLETON,
            ...store.scopes,
            ...configuration.scopes
        },
        mount: (0, di_1.mergeMount)(store.mount, configuration.mount),
        logger: {
            debug: false,
            level: env === core_1.Env.TEST ? "off" : "info",
            logRequest: true,
            ...store.logger,
            ...configuration.logger,
            jsonIndentation: process.env.NODE_ENV === core_1.Env.PROD ? 0 : 2
        }
    };
}
exports.getConfiguration = getConfiguration;
